src/rust/engine/serverset/src/retry.rs

```diff
@@ -18,28 +41,47 @@ impl<T: Clone + Send + Sync + 'static> Retry<T> {
   >(
     &self,
     f: F,
-    times: usize,
+    retry_params: RetryParameters,
   ) -> impl Future<Item = Value, Error = String> {
     let serverset = self.0.clone();
+    let RetryParameters { times, description } = retry_params;
     futures::future::loop_fn(0_usize, move |i| {
       let serverset = serverset.clone();
+      let description = description.clone();
+      let description2 = description.clone();
       let f = f.clone();
       serverset
         .next()
         .and_then(move |(server, token)| {
           futures::future::ok(server).and_then(f).then(move |result| {
             let health = match &result {
               &Ok(_) => Health::Healthy,
-              &Err(_) => Health::Unhealthy,
+              Err(err) => {
+                debug!(
+                  "({}) Attempt {} for {} failed with error {:?}",
+                  now(),
+                  i,
+                  description,
+                  err
+                );
+                Health::Unhealthy
+              }
             };
             serverset.report_health(token, health);
+            debug!("({}) Attempt {} for {} succeeded!", now(), i, description);
             result
           })
         })
         .map(futures::future::Loop::Break)
         .or_else(move |err| {
```

Comment:
I would probably fold this into the `and_then` above, rather than double-cloning `description`, but also that's probably a bit of a micro-optimisation